compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/PropertyReferenceLowering.kt

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.lower
+
+import org.jetbrains.kotlin.backend.common.ClassLoweringPass
+import org.jetbrains.kotlin.backend.common.IrElementTransformerVoidWithContext
+import org.jetbrains.kotlin.backend.common.ir.copyTo
+import org.jetbrains.kotlin.backend.common.ir.createImplicitParameterDeclarationWithWrappedDescriptor
+import org.jetbrains.kotlin.backend.common.lower.createIrBuilder
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.backend.jvm.JvmLoweredDeclarationOrigin
+import org.jetbrains.kotlin.descriptors.*
+import org.jetbrains.kotlin.ir.builders.*
+import org.jetbrains.kotlin.ir.builders.declarations.*
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.expressions.impl.*
+import org.jetbrains.kotlin.ir.symbols.*
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.types.impl.*
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.ir.visitors.*
+import org.jetbrains.kotlin.load.java.JvmAbi
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.types.Variance
+
+internal val propertyReferencePhase = makeIrFilePhase(
+    ::PropertyReferenceLowering,
+    name = "PropertyReference",
+    description = "Construct KProperty instances returned by expressions such as A::x and A()::x"
+)
+
+internal class PropertyReferenceLowering(val context: JvmBackendContext) : ClassLoweringPass {
+    // TODO: join IrLocalDelegatedPropertyReference and IrPropertyReference via the class hierarchy?
+    private val IrMemberAccessExpression.getter: IrFunctionSymbol?
+        get() = (this as? IrPropertyReference)?.getter ?: (this as? IrLocalDelegatedPropertyReference)?.getter
+
+    private val IrMemberAccessExpression.setter: IrFunctionSymbol?
+        get() = (this as? IrPropertyReference)?.setter ?: (this as? IrLocalDelegatedPropertyReference)?.setter
+
+    private val IrMemberAccessExpression.field: IrFieldSymbol?
+        get() = (this as? IrPropertyReference)?.field
+
+    private val IrMemberAccessExpression.signature: String
+        // Plain Java fields do not have a getter, but can be referenced nonetheless.
+        get() = getter?.owner?.let { context.state.typeMapper.mapSignatureSkipGeneric(it.descriptor).toString() }
+            ?: TODO("plain Java field signature")
+
+    private val IrMemberAccessExpression.symbol: IrSymbol
+        get() = getter?.owner?.symbol ?: field!!.owner.symbol
+
+    private fun IrFunction.visibleFrom(scope: Scope): Boolean =
```

Comment:
I think we should avoid doing complicated logic like this in the compiler backend. It's the frontend's job to resolve everything down to the specific types that the backend should use in code generation, including whether to use `KMutableProperty` or `KProperty` as a property reference superclass.

Currently the frontend already infers this information in `DoubleColonExpressionResolver.createKCallableTypeForReference` and stores it into `BindingContext.VARIABLE` binding context slice (at `DoubleColonExpressionResolver.bindPropertyReference`). It seems that `psi2ir` loses this information when transforming `KtCallableReferenceExpression` to `IrCallableReference` in `ReflectionReferencesGenerator.generateCallableReference`. Could you please look into obtaining that information there and changing the created `IrPropertyReference` to have no setter if it's not visible? I.e. get the "synthetic" local variable via `getOrFail(BindingContext.VARIABLE, ktCallableReference)` and only generate a setter in the property reference if that variable is mutable.

Reply:
Seems like the type is the *only* place where the frontend records mutability, which is not exactly convenient as the information about subtyping is not available afterwards, and checking against specific types seems like a hack anyway. Would it be ok to modify the frontend to also set `isVar` of the synthetic local variable accordingly?

Reply:
I think it should be fine. That local variable is only needed for backends anyway, so I hope no other code relies on it being a `val` right now.